tutorial/lesson_18_parallel_associative_reductions.cpp

```diff
@@ -0,0 +1,325 @@
+        // free, the intermediate reduction is now parallelizable across those
+        // dimensions. All reduction variables not in the list are removed from
+        // the original function and "lifted" to the intermediate.
+
+        // To generate the same code as the manually-factored version, we do the
+        // following:
+        Func intermediate = histogram.update().rfactor({{r.y, y}});
+        // We pass {r.y, y} as the argument to rfactor to make the histogram
+        // parallelizable across the y dimension, similar to the manually-factored
+        // version.
+        intermediate.compute_root().update().parallel(y);
+
+        // Since there is only one pair passed to rfactor, we could also write
+        // the previous rfactor this way:
+        // Func intermediate = histogram.update().rfactor(r.y, y);
+
+        // Vectorize the initializations to make things faster.
+        intermediate.vectorize(x, 8);
+        histogram.vectorize(x, 8);
+
+        // It is important to note that rfactor (or reduction factorization in
+        // general) only works for associative reductions. Associative reductions
+        // have the nice property that their results are the same no matter how
+        // the computation is grouped (i.e. split into chunks). If rfactor can't
+        // prove the associativity of a reduction, it will throw an error.
+
+        Buffer<int> halide_result = histogram.realize(8);
+
+        // See figures/lesson_18_hist_rfactor_par.gif for a visualization of
+        // what this did.
+
+        // The equivalent C is:
+        int c_intm[input.height()][8];
+        for (int y = 0; y < input.height(); y++) {
+            for (int x = 0; x < 8; x++) {
+                c_intm[y][x] = 0;
+            }
+        }
+        /* parallel */ for (int y = 0; y < input.height(); y++) {
+            for (int r_x = 0; r_x < input.width(); r_x++) {
+                c_intm[y][input(r_x, y) % 8] += 1;
+            }
+        }
+
+        int c_result[8];
+        for (int x = 0; x < 8; x++) {
+            c_result[x] = 0;
+        }
+        for (int x = 0; x < 8; x++) {
+            for (int r_y = 0; r_y < input.height(); r_y++) {
+                c_result[x] += c_intm[r_y][x];
+            }
+        }
+
+        // Check the answers agree:
+        for (int x = 0; x < 8; x++) {
+            if (c_result[x] != halide_result(x)) {
+                printf("halide_result(%d) = %d instead of %d\n",
+                       x, halide_result(x), c_result[x]);
+                return -1;
+            }
+        }
+    }
+
+    {
+        // Since we can factor associative reductions through the scheduling
+        // directive 'rfactor', it is really easy to explore various factorization
+        // strategies. Given the same serial histogram code:
+        Func histogram("hist_rfactor_vec");
+        histogram(x) = 0;
+        RDom r(0, input.width(), 0, input.height());
+        histogram(input(r.x, r.y) % 8) += 1;
```

Comment:
Use div to shrink the range, not mod. It's weird to take a histogram of the last few bits.